### PR TITLE
[ENG-373] Create periodic task that checks and sets stuck DOIs

### PR DESCRIPTION
**Pull request by Johnetordoff**
## Purpose

We've had several DOI confirmation emails that may have been lost during downtime. This PR will check daily to see if any DOI confirmations have been missed.   

## Changes

- add periodic script and tests

## QA Notes

Preprints that are missing DOI confirmation (there's no link to the preprint's DOI) - should appear at 11 PM.   

In summary, if we have published preprints, without the `preprint_doi_created` set, we filter crossref's API for these DOI's.  If they exist, we set the DOI in our db.  Otherwise, we email the help desk.

## Documentation

Not user facing, no docs.

## Side Effects

None that I know of.

## Ticket

https://openscience.atlassian.net/browse/ENG-373

### CODE REVIEW

File: scripts/periodic/check_crossref_dois.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+                                                      is_published=True)
+
+    pending_dois = []
+
+    for preprint in preprints_with_pending_dois:
+        prefix = preprint.provider.doi_prefix
+        pending_dois.append('doi:{}'.format(settings.DOI_FORMAT.format(prefix=prefix, guid=preprint._id)))
+
+    url = '{}/works?filter={}'.format(settings.CROSSREF_JSON_API_URL, ','.join(pending_dois))
+    preprints = requests.get(url).json()['message']['items']
```
Comment (brianjgeiger) [inline scripts/periodic/check_crossref_dois.py]: What happens when this fails? Does the script just not run for the day? If so, maybe include a retry mechanism (and test for this condition).

File: scripts/periodic/check_crossref_dois.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+            to_addr=settings.OSF_SUPPORT_EMAIL,
+            mail=mails.CROSSREF_DOIS_PENDING,
+            pending_doi_count=preprints_with_pending_dois.count(),
+            email_content=content,
+        )
+        logger.error('There were {} stuck registrations for CrossRef, email sent to help desk'.format(preprints_with_pending_dois.count()))
+
+
+@celery_app.task(name='scripts.periodic.check_crossref_dois')
+def run_main(dry_run=True):
```
[the remark this answers is not in the document]
Comment (Johnetordoff) [reply scripts/periodic/check_crossref_dois.py]: It's periodic and all other periodic things that run are written as scripts, so this is more consistent. My understanding is that management commands are always meant to be manually triggered, but I'm unsure,

File: scripts/periodic/check_crossref_dois.py (hunk 1 above)
Comment (Johnetordoff) [reply scripts/periodic/check_crossref_dois.py]: A retry mechanism is built in to requests, but I will improve the exception handling here to log and throw the error better.

File: website/settings/defaults.py [hunk 3]
```diff
@@ -578,6 +585,10 @@ class CeleryConfig:
                 'task': 'scripts.generate_sitemap',
                 'schedule': crontab(minute=0, hour=5),  # Daily 12:00 a.m.
             },
+            'check_crossref_doi': {
+                'task': 'scripts.periodic.check_crossref_dois',
+                'schedule': crontab(minute=0, hour=5),  # Daily 12:00 a.m.
```
[the remark this answers is not in the document]
Comment (Johnetordoff) [reply website/settings/defaults.py]: Checked with Matt he agrees, I changed it to 11:00pm just because we have things running at 1am and 2am as well.

File: scripts/periodic/check_crossref_dois.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (brianjgeiger) [reply scripts/periodic/check_crossref_dois.py]: Here's a couple of examples of how I'm handling this in upcoming PRs.

https://github.com/CenterForOpenScience/osf.io/pull/9050/files#diff-c39b88b6fef6c8476d857a5c9da960d6R449

https://github.com/CenterForOpenScience/osf.io/pull/9075/files#diff-ea015ffe216ce38d14dea62852418f34R122

[the remark this answers is not in the document]
Comment (brianjgeiger) [reply scripts/periodic/check_crossref_dois.py]: The general idea is, just because this will be run periodically, doesn't mean it should only *have* to be run periodically. If crossref is down at 11 and we want to run it again first thing in the morning, having it as a management command makes it easy to do.

File: website/settings/defaults.py [hunk 4]
```diff
@@ -578,6 +585,10 @@ class CeleryConfig:
                 'task': 'scripts.generate_sitemap',
                 'schedule': crontab(minute=0, hour=5),  # Daily 12:00 a.m.
             },
+            'check_crossref_doi': {
+                'task': 'osf.management.commands.check_crossref_dois',
```
Comment (brianjgeiger) [inline website/settings/defaults.py]:
```suggestion
                'task': 'management.commands.check_crossref_dois',
```

File: osf/management/commands/check_crossref_dois.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    check_crossref_dois()
+    report_stuck_dois()
+
+
+class Command(BaseCommand):
+    help = '''Checks if we've missed any Crossref DOI confirmation emails. '''
+
+    # Management command handler
+    def handle(self, *args, **options):
+        check_crossref_dois()
```
Comment (brianjgeiger) [inline osf/management/commands/check_crossref_dois.py]: If you wanted, you could just call `main()` here.

File: osf/management/commands/check_crossref_dois.py [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+
+@celery_app.task(name='management.commands.check_crossref_dois')
+def main():
+    check_crossref_dois()
+    report_stuck_dois()
+
+
+class Command(BaseCommand):
+    help = '''Checks if we've missed any Crossref DOI confirmation emails. '''
+
```
Comment (brianjgeiger) [inline osf/management/commands/check_crossref_dois.py]: A `dry_run` option would probably be really useful since we can't run this on prod data without emailing people.

File: website/settings/defaults.py (hunk 4 above)
Comment (brianjgeiger) [reply website/settings/defaults.py]: I believe it's using the name you provided to the task decorator and not the actual path to the script, so I think this will work better.

File: osf/management/commands/check_crossref_dois.py (hunk 6 above)
Comment (brianjgeiger) [reply osf/management/commands/check_crossref_dois.py]: Would it be at all helpful to be able to optionally disable one of the two functions? Are there times when someone might want to run one but not the other?

Comment (Johnetordoff) [reply osf/management/commands/check_crossref_dois.py]: You want to run these at one after another. You wouldn't want it to email you then have it set a pending DOI and make the email irreverent.

File: osf/management/commands/check_crossref_dois.py [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+            mails.send_mail(
+                to_addr=settings.OSF_SUPPORT_EMAIL,
+                mail=mails.CROSSREF_DOIS_PENDING,
+                pending_doi_count=preprints_with_pending_dois.count(),
+                email_content=content,
+            )
+        else:
+            logger.info('DRY RUN')
+
+        logger.error('There were {} stuck registrations for CrossRef, email sent to help desk'.format(preprints_with_pending_dois.count()))
```
Comment (pattisdr) [inline osf/management/commands/check_crossref_dois.py]: maybe `logger.info` instead of `logger.error`? This makes me think that something went wrong.

File: osf/management/commands/check_crossref_dois.py [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+
+    if preprints_with_pending_dois:
+        guids = ', '.join(preprints_with_pending_dois.values_list('guids___id', flat=True))
+        content = 'DOIs for the following preprints have been pending at least {} days: {}'.format(time_since_published.days, guids)
+        if not dry_run:
+            mails.send_mail(
+                to_addr=settings.OSF_SUPPORT_EMAIL,
+                mail=mails.CROSSREF_DOIS_PENDING,
+                pending_doi_count=preprints_with_pending_dois.count(),
+                email_content=content,
```
Comment (pattisdr) [inline osf/management/commands/check_crossref_dois.py]: You need to define a mako template `crossref_doi_pending.html.mako` with the contents of the email.  I'd go ahead and move `content` to the mako template, and just pass in the guids and the days variable to that template.

File: website/mails/mails.py [hunk 9]
```diff
@@ -435,6 +435,11 @@ def get_english_article(word):
     subject='There was an error creating a DOI for preprint(s). batch_id: ${batch_id}'
 )
 
+CROSSREF_DOIS_PENDING = Mail(
+    'crossref_doi_pending',
+    subject='There are ${pending_doi_count} pending.'
```
Comment (pattisdr) [inline website/mails/mails.py]: Could you make the subject of this email a little more clear? nothing will be sent with the word crossref in it, it will just say `There are x pending`.

File: osf/management/commands/check_crossref_dois.py [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+        else:
+            logger.info('DRY RUN')
+
+
+def report_stuck_dois(dry_run=True):
+    time_since_published = timedelta(days=settings.DAYS_CROSSREF_DOIS_MUST_BE_STUCK_BEFORE_EMAIL)
+
+    preprints_with_pending_dois = Preprint.objects.filter(preprint_doi_created__isnull=True,
+                                                          is_published=True,
+                                                          date_published__lt=timezone.now() - time_since_published)
```
Comment (pattisdr) [inline osf/management/commands/check_crossref_dois.py]: note to self - if no DOI's match from querying https://api.crossref.org/, and it's been published for five days, it will show up in this query - these guids are sent to the help desk.

File: osf/management/commands/check_crossref_dois.py [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+import django
+django.setup()
+
+from osf.models import Preprint
+
+logger = logging.getLogger(__name__)
+logging.basicConfig(level=logging.INFO)
+
+
+def check_crossref_dois(dry_run=True):
```
Comment (mfraezz) [inline osf/management/commands/check_crossref_dois.py]: Minor: Why default to `True` here but `False` in main?

File: osf/management/commands/check_crossref_dois.py [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+    """
+    This script is to check for any DOI confirmation messages we may have missed during downtime and alert admins to any
+    DOIs that have been pending for X number of days.
+    :param dry_run:
+    :return:
+    """
+    # Create one enormous url to check if all our pending crossref DOIs are good, then set all identifers
+    preprints_with_pending_dois = Preprint.objects.filter(preprint_doi_created__isnull=True,
```
Comment (mfraezz) [inline osf/management/commands/check_crossref_dois.py]: Should this also include a "stuck" timedelta filter?

File: osf/management/commands/check_crossref_dois.py [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+        logger.error('Could not contact crossref to check for DOIs, response returned with exception {}'.format(exc))
+        raise exc
+
+    preprints = resp.json()['message']['items']
+
+    for preprint in preprints:
+        guid = preprint['DOI'].split('/')[-1]
+        pending_preprint = preprints_with_pending_dois.get(guids___id=guid)
+        if not dry_run:
+            pending_preprint.set_identifier_values(preprint['DOI'])
```
Comment (mfraezz) [inline osf/management/commands/check_crossref_dois.py]: Not certain where to put this, but here seems most relevant --

What happens when the problem isn't us missing Crossref's issuance email, but them missing our issuance request? Do we already handle that case anywhere?

File: website/settings/defaults.py [hunk 14]
```diff
@@ -334,6 +334,13 @@ def parent_dir(path):
 ECSARXIV_CROSSREF_USERNAME = None
 ECSARXIV_CROSSREF_PASSWORD = None
 
+# if our DOIs cannot be comfirmed after X amount of days email the admin
+DAYS_CROSSREF_DOIS_MUST_BE_STUCK_BEFORE_EMAIL = 5
```
Comment (mfraezz) [inline website/settings/defaults.py]: Minor: I believe we say that it may take up to 24 hours for Crossref to issue a DOI; `2` might be a more reasonable value here.

File: osf/management/commands/check_crossref_dois.py (hunk 12 above)
Comment (Johnetordoff) [reply osf/management/commands/check_crossref_dois.py]: I discussed with product, they don't have a problem with DOIs that have been "stuck" for an hour or however long getting set here. Seems extraneous, could potential cause problems if we have to change the time window for being "stuck".

File: osf/management/commands/check_crossref_dois.py (hunk 13 above)
Comment (Johnetordoff) [reply osf/management/commands/check_crossref_dois.py]: Not sure, to best of my knowledge that has never happened, it is a celery task I believe so there's like a retry in there somewhere, but good point.

File: osf/management/commands/check_crossref_dois.py (hunk 11 above)
Comment (Johnetordoff) [reply osf/management/commands/check_crossref_dois.py]: Yeah, this makes no sense, It should default to true everywhere.

File: osf/management/commands/check_crossref_dois.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (Johnetordoff) [reply osf/management/commands/check_crossref_dois.py]: If I understand to correctly, there's no real need to keep DOI specific info like that, Crossref keeps that info in fine grained detail in it's API and it will reliably be there if we need it to debug.

Second, 👍

File: osf/management/commands/check_crossref_dois.py (hunk 12 above)
Comment (mfraezz) [reply osf/management/commands/check_crossref_dois.py]: Maybe a better question would be "does it harm anything if we include preprints made 10 seconds before this script runs?"

If Crossref just ignores it, I suppose it doesn't matter much, but what if this fetches and sets it on our side before the email they send is delivered and parsed? Does that part have appropriate error handling if the preprint already has a DOI?

Comment (Johnetordoff) [reply osf/management/commands/check_crossref_dois.py]: Good point, I'll add this in

File: scripts/tests/test_check_crossref_dois.py [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+import json
+from datetime import timedelta
+import responses
+HERE = os.path.dirname(os.path.abspath(__file__))
+
+
+from osf_tests.factories import PreprintFactory
+from website import settings, mails
+
+from scripts.periodic.check_crossref_dois import check_crossref_dois, report_stuck_dois
```
Comment (mfraezz) [inline scripts/tests/test_check_crossref_dois.py]: Import error here.

File: scripts/tests/test_check_crossref_dois.py [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+
+
+@pytest.mark.django_db
+class TestCheckCrossrefDOIs:
```
Comment (mfraezz) [inline scripts/tests/test_check_crossref_dois.py]: These tests don't work even after fixing the import error; clearly they aren't being run by TCI. Given that we plan on running this cron, it would probably be good to have these tests actually run, unlike the rest of the `scripts` directory. It might make sense to put them in an `osf_tests/test_management_commands.py`.
```
================================================================================================================================================================================= FAILURES =================================================================================================================================================================================
_______________________________________________________________________________________________________________________________________________________________ TestCheckCrossrefDOIs.test_report_stuck_dois _______________________________________________________________________________________________________________________________________________________________
scripts/tests/test_check_crossref_dois.py:74: in test_report_stuck_dois
    to_addr=settings.OSF_SUPPORT_EMAIL)
../../.pyenv/versions/2.7.10/envs/osf/lib/python2.7/site-packages/mock/mock.py:937: in assert_called_with
    six.raise_from(AssertionError(_error_message(cause)), cause)
../../.pyenv/versions/2.7.10/envs/osf/lib/python2.7/site-packages/six.py:737: in raise_from
    raise value
E   AssertionError: Expected call: send_mail(email_content='DOIs for the following preprints have been pending at least 2 days: guid0', mail=<website.mails.mails.Mail object at 0x113ce0610>, pending_doi_count=1, to_addr='fake-support@osf.io')
E   Actual call: send_mail(guids=u'guid0', mail=<website.mails.mails.Mail object at 0x113ce0610>, pending_doi_count=1, time_since_published=2, to_addr='fake-support@osf.io')
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------- Captured log call -----------------------------------------------------------------------------------------------------------------------------------------------------------------------------
check_crossref_dois.py      85 INFO     There were 1 stuck registrations for CrossRef, email sent to help desk
______________________________________________________________________________________________________________________________________________________________ TestCheckCrossrefDOIs.test_check_crossref_dois ______________________________________________________________________________________________________________________________________________________________
<string>:3: in wrapper
    ???
scripts/tests/test_check_crossref_dois.py:62: in test_check_crossref_dois
    assert stuck_preprint.identifiers.count() == 1
E   assert 0 == 1
E    +  where 0 = <bound method GenericRelatedObjectManager.count of <django.contrib.contenttypes.fields.GenericRelatedObjectManager object at 0x11709a390>>()
E    +    where <bound method GenericRelatedObjectManager.count of <django.contrib.contenttypes.fields.GenericRelatedObjectManager object at 0x11709a390>> = <django.contrib.contenttypes.fields.GenericRelatedObjectManager object at 0x11709a390>.count
E    +      where <django.contrib.contenttypes.fields.GenericRelatedObjectManager object at 0x11709a390> = <Preprint: Organized zero tolerance help-desk (published preprint) (guid=ebgfx)>.identifiers
```

File: website/settings/defaults.py [hunk 17]
```diff
@@ -588,6 +595,10 @@ class CeleryConfig:
                 'task': 'scripts.generate_sitemap',
                 'schedule': crontab(minute=0, hour=5),  # Daily 12:00 a.m.
             },
+            'check_crossref_doi': {
+                'task': 'management.commands.check_crossref_dois',
```
Comment (mfraezz) [inline website/settings/defaults.py]: Forgot one thing -- this needs to be added to `CeleryConfig.imports`

Comment (Johnetordoff) [reply website/settings/defaults.py]: 👍

File: scripts/tests/test_check_crossref_dois.py (hunk 16 above)
Comment (Johnetordoff) [reply scripts/tests/test_check_crossref_dois.py]: :(

File: .travis.yml [hunk 18]
```diff
@@ -5,6 +5,8 @@ language: python
 python:
     - "2.7"
 
+dist: 'trusty'
```
Comment (mfraezz) [inline .travis.yml]: Mild scope creep, but it's fine.
